### PR TITLE
Add dvorak layout

### PR DESCRIPTION
**Pull request by wildeyedskies**

### CODE REVIEW

**Comment by UniversalSuperBox** [comment]
Just coming in from the high-level overview, `en-dv` implies that this is a new locale (like en-US and en-UK). I wonder if we should choose another naming convention for a layout that meets the same language requirements as another, but is a different layout.

@kugiigi, thoughts?

**Comment by wildeyedskies** [comment]
Any suggestions for what that convention should be? I didn't want to just do dv because that implies a separate locale, which isn't really the case.

**Comment by UniversalSuperBox** [comment]
Maybe an underscore rather than a dash?

**Comment by UniversalSuperBox** [comment]
...or @. I wonder if that would mess up the build system.

**Comment by UniversalSuperBox** [comment]
Yes, @dobey recommended `en_US@layout`. That seems sane. Since this is not country-specific, simply `en@layout` makes sense, if the keyboard can still load it.

**Comment by Fuseteam** [comment]
i tried en@flick directly on the device and it worked fine

**Comment by wildeyedskies** [comment]
I will change the locale code to en@dv and rebuild the package.

**Comment by kugiigi** [comment]
I've just read this and learned about Dvorak layout but anyway it seems like this has been resolved 😅

I wonder though if this convention would be the thing forward and perhaps the start of supporting different layout per language like in Android? Someone else is trying to create a flick-style english layout.

**Comment by wildeyedskies** [comment]
@UniversalSuperBox I've migrated to en@dv. Does anything else need to be updated?

**Comment by UniversalSuperBox** [comment]
I'm trying to figure out why, but the language plugin still compiles as `libenplugin.so`, causing the keyboard to fail to load it.

File: plugins/en@dv/src/src.pro
Comment (UniversalSuperBox) [inline plugins/en@dv/src/src.pro]: Change this to `en@dvplugin` to name the `.so` file correctly.

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
Aha, found the issue!

**Comment by wildeyedskies** [comment]
Try now

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
I've found two issues:

1. Loading the layout crashes the keyboard
1. The layout is called 'en@dv' rather than 'English (Dvorak)' in the settings

This now crashes when loading the keyboard:

```
WARNING: Did not find a dictionary for "en@dv"  - checking for  "en"
[DatabaseConnector] Unable to create/open database: /usr/share/maliit/plugins/com/ubuntu/lib/en@dv/database_en@dv.db
terminate called after throwing an instance of 'SqliteDatabaseConnector::SqliteDatabaseConnectorException'
  what():  unable to open database file
```

This is an interesting error, it appears it should be falling back to the `en` plugin. There are a couple of options here:

1. Edit the packaging so the dvorak keyboard depends on the normal English keyboard and `en@dv/database_en@dv.db` is a symlink to `../en/database_en.db`.
1. Make sure that fallback code it mentions works correctly

I'd prefer option 2, if possible. Or any other options you see.

Do you have an Ubuntu Touch device that you're using to test these changes, perchance? You can install the keyboard by running `sudo ubports-qa install PR-86` whenever you like. If it complains about being out of space, run `sudo mount / -o remount,rw` followed by `sudo apt clean` followed by `sudo mount -t tmpfs tmpfs /var/cache/apt/archives`, then try again.

**Comment by wildeyedskies** [comment]
My device got a bit messed up from some on-device prototyping, but I will see if I can get it to run the tests properly. The database not loading seems to be a naming issue. It should be loading the English database.

**Comment by UniversalSuperBox** [comment]
Yes, if the database was renamed then it would be fine. The problem is, you're using the same database as the English version and it's about 5MB. That doesn't sound like a lot, but it's pretty huge for Ubuntu Touch packages... and it's all wasted space. It'd be better to reuse the same English database rather than shipping a second copy.

**Comment by wildeyedskies** [comment]
I agree, we should not duplicate the database if possible. I think it makes more sense to have the Dvorak layout in the same package as the regular English layout, rather than a separate package with a dependency.

**Comment by Fuseteam** [comment]
maybe this is a good opportunity to find a way to separate language and layout

**Comment by wildeyedskies** [comment]
@Fuseteam I think that's a good idea, but I'm not quite sure how to go about it. Should the layouts still be in separate folders or should it be something like en/en@dv?

**Comment by tallero** [comment]
> I agree, we should not duplicate the database if possible. I think it makes more sense to have the Dvorak layout in the same package as the regular English layout, rather than a separate package with a dependency.

About the old decision (because I would like to have a working keyboard to set a better package for the keyboard):
I tried setting the `database-en@dv.db` as a link, but
- or does the `qinstall` does not install symlinks. 
- or the files are not picked.

Can you check the new changes in #90 ?

**Comment by Fuseteam** [comment]
@zoenb i was thinking of a separate folder where the layouts reside while the  databases stay where they are
so that layout and database are from different folders
the sheme 'language@layout' could then be used to read the correct language and layout

**Comment by wildeyedskies** [comment]
@UniversalSuperBox @Fuseteam Would it be reasonable to change the way the westernlanguage plugin handles finding the database? Currently it just looks for `database_<locale>.db` in the plugin's directory. I'd rather avoid copying the database or using symlinks. I have the following suggestions.

1. Allow for the specification of the database file path possibly via the json file or a separate config file for the plugin, which can then be read by setLanguage in the spellpredictworker.cpp file. This would be the most flexible solution but it would also be the most complex, requiring changes to every plugin
2. When looking up the database drop everything after `@`. This would be far simpler to implement and would be consistent if we have additional custom layouts to add in the future, but it is less flexible.

Do either of these solutions sound good? If so, which one would be better?

**Comment by Fuseteam** [comment]
@zoenb what makes 1 more flexible than 2?
i think option 2 sounds more reasonable

i'll leave the final call to @UniversalSuperBox tho

**Comment by UniversalSuperBox** [comment]
Problem with 2 is that the database is loaded from the plugin's folder, the same directory where the QML file is located. Unless we start putting multiple layouts in the same folder (as you mentioned), which could work for now. However it'd mean redoing the plugin loading model a bit. Well, it seems like any change we made would mean redoing it a bit.

How deep have you gone into the maliit/ubuntu-keyboard codebase? I've only made a few patches here and there, I'm by no means an expert. You probably have more knowledge that lets you decide the best course of action.

Very long term, I'd like to be able to install layouts or dictionaries from clicks or another packaging format. Short term, I'd say do what's easiest to satisfy the constraints ("Don't copy the database") and don't sweat the rest, or come back to fix it later. This PR did start as a way to add a dvorak layout, after all.

Thank you very much, by the way. I don't think I say that to PR authors enough.

**Comment by wildeyedskies** [comment]
I think solution 2 is what I'm going to go with. I'm looking to modify the database loading code which lives in the westernlanguagesupport plugin such that it'll ignore all characters after a `@` when it looks for a database. I *think* this is how it should all work, but I'm very new to C++ and I really know next to nothing about maliit itself.

Currently I'm trying to get the build to run without a database, but its failing at dh_install with files missing. I'm not sure where the database files are specified that's causing this to fail.
```
dh_install: ubuntu-keyboard-english-dvorak missing files: usr/share/maliit/plugins/com/ubuntu/lib/en@dv/
dh_install: missing files, aborting
debian/rules:37: recipe for target 'override_dh_install' failed
```

**Comment by tallero** [comment]
OT: I noticed that many projects using `maalit` use different directory build layouts and format. The reason you did not have a patch like this from me in 2016 was that I could not reuse sailfish's dvorak qml layout here.

If it is useful to know, I also noticed from #90 that `qinstall` skips symlinks.

**Comment by UniversalSuperBox** [comment]
@zoenb, check `debian/ubuntu-keyboard-english-dvorak.install`. This tells dh_install to install those files. If you don't have a `usr/share/maliit/plugins/com/ubuntu/lib/en@dv/` any more, the .install file can be removed. It probably needs to be changed, at least.

**Comment by zoenb** [comment]
I still have `usr/share/maliit/plugins/com/ubuntu/lib/en@dv/` for the layout, it just doesn't have the database files anymore.

**Comment by UniversalSuperBox** [comment]
Awesome, now we're cooking! Your layout uses just 53K on disk.

For the layout, is this what you expected it to look like? Do note that you can add extra keys to each row if you need to fit more keys. Q and Z surrounding the spacebar seems a little strange compared to other OSK Dvorak layouts I've seen.

Dvorak keyboard layout

Ignore the spacebar saying "En@dv", that was fixed by a reboot.

**Comment by wildeyedskies** [comment]
I based the layout on Andorid's dvorak keyboard, and it has the Q and Z keys in those positions.

**Comment by UniversalSuperBox** [comment]
Ah interesting. I have no idea how I missed that when searching for 'Dvorak on-screen keyboard'.

**Comment by UniversalSuperBox** [comment]
I notice a few keys are different from the English keyboard here. In particular, I note that the "D" key is missing its extensions. It's *extremely* unlikely that these characters would be typed in English, however I'd prefer if the new layout followed the convention set by its base layout. That is, if a key cannot be accessed from another location on this layout, just copy the key from the [default English layout](https://github.com/ubports/keyboard-component/blob/xenial/plugins/en/qml/Keyboard_en.qml). Remember to repeat that change in all of your new layouts too.

If you have a good reason not to follow convention, I'll probably agree with you though. What do you think?

**Comment by wildeyedskies** [comment]
I'll add the D key accents tomorrow. Is there anything else we need before this is good to merge?

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
Yes. T is also missing its accents from the EN keyboard. I believe all these characters are on the English keyboard so that people who do not have a keyboard for their language can still use Ubuntu Touch, and also because English is the fallback for those.

**Comment by tallero** [comment]
I am using @zoenb layout from last week and it is awesome.

Just to propose, what do you guys think about adding arrow keys in the long press popup on a cross layout or putting them on a third level? I remember this were feasible when playing with maliit on sailfish. There should be keycodes for them somewhere on mer repositories.

I really became addicted to having arrow keys in android navbar (lineageos specific) and it would be really useful to have them everywhere, not only in the terminal.

EDIT: It is just a proposition, let's first merge this request into master.

**Comment by UniversalSuperBox** [comment]
@tallero, please file another request for changes. This will be lost here. :)

**Comment by Fuseteam** [comment]
sounds like a request for an interaction layer xD

**Comment by Fuseteam** [comment]
@UniversalSuperBox is T the only one left that's missing its accents?

**Comment by wildeyedskies** [comment]
Alright, that *should* be all the missing accents from the en keyboard.

**Review by UniversalSuperBox — APPROVED** [review]
Everything seems to be in order here. Nice work, and huge thanks! I'll have to get this into the metapackage so it's installed in Ubuntu Touch by default.

**Comment by UniversalSuperBox** [comment]
@zoenb, is there any specific way you would like to be credited in our changelog for our next release? Would you prefer your username, or another name...?

**Comment by wildeyedskies** [comment]
Glad to see it get merged. Just zoenb is fine for the changelog. Thanks for asking!

**Comment by UniversalSuperBox** [comment]
No problem. zoenb it is! Thank you very much. Have you checked out any other parts of Ubuntu Touch, maybe there's something I could point you in the right direction on.

**Comment by wildeyedskies** [comment]
I've poked through the code a little bit. Do you have a list of work for centralized anywhere or is it just in the individual project issue trackers?

**Comment by Fuseteam** [comment]
Congrats!

**Comment by UniversalSuperBox** [comment]
It depends on what you're looking for. The issue tracker for the system is https://github.com/ubports/ubuntu-touch/issues, but if we know that an issue only affects a single component we may move it to that component's repository. Core apps are moving over to https://gitlab.com/ubports. Overarching projects, like stable system updates, are tracked at https://github.com/orgs/ubports/projects. We'd love to hear from you in any of our chatrooms (#ubports on freenode, @ubports on Telegram, #ubports:matrix.org on the Matrix network) or on the forum at forums.ubports.com, if you'd like.

**Comment by ziggutas** [comment]
Devices: Nexus 7 (flo), BQ E4.5 ubuntu edition. Both on rc (2019-W42)
English Dvorak keyboard layout works on these two devices. Also the special characters for the D and T on the English layout. Always wondered what those were! Feel more acquainted with my keyboard now :D

**Pull request merged**
